#01. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path_name = r'/Users/geetika/Documents/Python/Instacart Basket Analysis'

In [3]:
# Importing customer data dataset
customer_data = pd.read_csv(os.path.join(path_name, '02_Data', 'Original Data', 'customers.csv'))

In [4]:
customer_data.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
customer_data.shape

(206209, 10)

In [6]:
customer_data.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

From the descriptive statistics, we can see that
1. the customer ages range from 18 to 81 - seems likely considering human lifespan,
2. the n_dependents range from 0 to 3, which is also not unusual,
3. the customer income has a very wide range with the mimimum being 25903 and the maximum being 593901. But if the max is truly an outlier and therefore needs to be exluded, will be determined by the questions that need to be answered.

#02. Data Wrangling

Dropping columns - All columns are relavent, therefore all columns are retained

Renaming columns - The column names are inconsistent in terms of case and spelling. They are corrected.

Changing Data types - The data type of column "date_joined" has been changed from object to <M8[ns].

Transposing Data - The table is intuitive to read. So, no trasposition is necessary

In [13]:
#customer_data = customer_data.rename(columns = {'Surnam' : 'Surname'}, inplace = False)

In [14]:
#customer_data = customer_data.rename(columns = {'n_dependants' : 'n_dependents'}, inplace = False)

- Renaming Columns

In [7]:
dict_rename = {'Surnam' : 'surname', 'Gender' : 'gender','STATE' : 'state', 'Age' : 'age','n_dependants' : 'number_of_dependents', 'fam_status' : 'marital_status'}

In [8]:
customer_data = customer_data.rename(columns = dict_rename, inplace = False)

In [9]:
customer_data.head()

user_id First Name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

- Data types

In [10]:
customer_data.dtypes

user_id                  int64
First Name              object
surname                 object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependents     int64
marital_status          object
income                   int64
dtype: object

In [11]:
customer_data['date_joined']
# month-day-year

0         1/1/2017
1         1/1/2017
2         1/1/2017
3         1/1/2017
4         1/1/2017
            ...   
206204    4/1/2020
206205    4/1/2020
206206    4/1/2020
206207    4/1/2020
206208    4/1/2020
Name: date_joined, Length: 206209, dtype: object

In [12]:
customer_data['date_joined'] = pd.to_datetime(customer_data['date_joined'], format = '%m/%d/%Y')

In [13]:
customer_data['date_joined'].dtype

dtype('<M8[ns]')

#03. Data Quality and Consistency checks

Finding and addressing mixed data types - fixed at data wrangling stage

Finding and addressing missing values - found in column "First name". No need for imputation or exclusion of missing values.

Finding and addressing duplicate records - No duplicates found.

- Addressing missing values

In [14]:
customer_data.isnull().sum()

user_id                     0
First Name              11259
surname                     0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependents        0
marital_status              0
income                      0
dtype: int64

Out of 206209 total observations, 11259 observations have missing first names. As all the remaining columns, and the "User_id" and "Surname" columns especially, are complete, it is not a problem if the "First Name" of some customers is missing. The user can still be identified with the help of the "User_id" and "Surname".

If this was a numeric data, the missing values could be either imputed with the mean or median of the column or filtered out.

- Finding Duplicate

In [15]:
customers_duplicates = customer_data[customer_data.duplicated()]

In [16]:
customers_duplicates

Empty DataFrame
Columns: [user_id, First Name, surname, gender, state, age, date_joined, number_of_dependents, marital_status, income]
Index: []

No duplicates found

- Checking for typos and inconsistencies in nomenclature

In [17]:
customer_data['gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: gender, dtype: int64

In [18]:
customer_data['state'].value_counts(dropna = False).sort_index()

Alabama                 4044
Alaska                  4044
Arizona                 4044
Arkansas                4044
California              4044
Colorado                4044
Connecticut             4044
Delaware                4044
District of Columbia    4044
Florida                 4044
Georgia                 4044
Hawaii                  4044
Idaho                   4044
Illinois                4044
Indiana                 4044
Iowa                    4044
Kansas                  4043
Kentucky                4043
Louisiana               4043
Maine                   4043
Maryland                4043
Massachusetts           4043
Michigan                4043
Minnesota               4043
Mississippi             4043
Missouri                4043
Montana                 4043
Nebraska                4043
Nevada                  4043
New Hampshire           4043
New Jersey              4043
New Mexico              4043
New York                4043
North Carolina          4043
North Dakota  

In [19]:
customer_data['marital_status'].value_counts(dropna = False)

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: marital_status, dtype: int64

#04. Combining customer data with prepared instacart data

In [20]:
# Importing the ords_prods_merge data set
ords_prods_merge = pd.read_pickle(os.path.join(path_name, '02_Data', 'Prepared Data', 'ords_prods_merge_E4.8.pkl'))

In [21]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  max_order  \
0         Soda        77              7     9.0   both         10   
1         Soda        77              7     9.0   both         10   
2         Soda        77              7     9.0   both         10   
3         Soda        77              7     9.0   both         10   
4         Soda        77              7     9.0   both         10   

   loyalty_flag  mean_price_of_products spending_flag  \
0  New customer                6.367797   Low spender   
1  New customer                6.367797   Low spender   
2  New customer                6.367797   Low spender   
3  New customer                6.367797   Low spender   
4  New customer                6.367797   Low spender   

   median_days_since_prior_order Regularity_of_customer  
0                           20.5  Non-frequent customer  
1                           20.5  Non-frequent customer  
2                           20.5  Non-frequent customer  
3                           20.5  Non-frequent customer  
4                           20.5  Non-frequent customer

In [22]:
ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [23]:
ords_prods_merge['user_id'].dtype

dtype('int64')

In [24]:
customer_data['user_id'].dtype

dtype('int64')

The data type of column "user_id" in both the dataframes is int64. So, we can use the merge() function.

In [25]:
ords_prods_merge.shape

(32434212, 19)

In [26]:
customer_data.shape

(206209, 10)

In [27]:
ords_prods_custs_merge = ords_prods_merge.merge(customer_data, on = 'user_id', indicator = True)

In [28]:
ords_prods_custs_merge.head(5)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  First Name  surname  gender    state age  date_joined  \
0         Soda  ...       Linda   Nguyen  Female  Alabama  31   2019-02-17   
1         Soda  ...       Linda   Nguyen  Female  Alabama  31   2019-02-17   
2         Soda  ...       Linda   Nguyen  Female  Alabama  31   2019-02-17   
3         Soda  ...       Linda   Nguyen  Female  Alabama  31   2019-02-17   
4         Soda  ...       Linda   Nguyen  Female  Alabama  31   2019-02-17   

  number_of_dependents  marital_status income _merge  
0                    3         married  40423   both  
1                    3         married  40423   both  
2                    3         married  40423   both  
3                    3         married  40423   both  
4                    3         married  40423   both  

[5 rows x 29 columns]

In [29]:
ords_prods_custs_merge['_merge'].value_counts()

both          32434212
left_only            0
right_only           0
Name: _merge, dtype: int64

In [30]:
ords_prods_custs_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'max_order', 'loyalty_flag',
       'mean_price_of_products', 'spending_flag',
       'median_days_since_prior_order', 'Regularity_of_customer', 'First Name',
       'surname', 'gender', 'state', 'age', 'date_joined',
       'number_of_dependents', 'marital_status', 'income', '_merge'],
      dtype='object')

In [33]:
ords_prods_custs_merge.shape

(32434212, 29)

In [31]:
# Descriptive Statistics of numeric columns
ords_prods_custs_merge.describe(include = [np.number])

order_id       user_id  order_number  orders_day_of_week  \
count  3.243421e+07  3.243421e+07  3.243421e+07        3.243421e+07   
mean   1.710748e+06  1.029374e+05  1.714192e+01        2.738810e+00   
std    9.872996e+05  5.946649e+04  1.753489e+01        2.090047e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.559450e+05  5.142000e+04  5.000000e+00        1.000000e+00   
50%    1.711049e+06  1.026125e+05  1.100000e+01        3.000000e+00   
75%    2.565510e+06  1.543910e+05  2.400000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01        6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.243421e+07            3.035613e+07  3.243421e+07   
mean        1.342500e+01            1.110416e+01  2.557778e+04   
std         4.246351e+00            8.778943e+00  1.409614e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.353500e+04   
50%         1.300000e+01            8.000000e+00  2.525600e+04   
75%         1.600000e+01            1.500000e+01  3.793500e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.243421e+07  3.243421e+07  3.243421e+07   3.243421e+07   
mean        8.351176e+00  5.896888e-01  7.121649e+01   9.922231e+00   
std         7.126688e+00  4.918902e-01  3.820429e+01   6.281190e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices     max_order  mean_price_of_products  \
count  3.243421e+07  3.243421e+07            3.243421e+07   
mean   1.198114e+01  3.305217e+01            1.198114e+01   
std    4.954311e+02  2.515449e+01            8.319087e+01   
min    1.000000e+00  3.000000e+00            1.000000e+00   
25%    4.200000e+00  1.300000e+01            7.391954e+00   
50%    7.400000e+00  2.600000e+01            7.829438e+00   
75%    1.130000e+01  4.700000e+01            8.258696e+00   
max    9.999900e+04  9.900000e+01            2.500542e+04   

       median_days_since_prior_order           age  number_of_dependents  \
count                   3.243421e+07  3.243421e+07          3.243421e+07   
mean                    1.039794e+01  4.946521e+01          1.501872e+00   
std                     7.131230e+00  1.848553e+01          1.118871e+00   
min                     0.000000e+00  1.800000e+01          0.000000e+00   
25%                     6.000000e+00  3.300000e+01          1.000000e+00   
50%                     8.000000e+00  4.900000e+01          2.000000e+00   
75%                     1.300000e+01  6.500000e+01          3.000000e+00   
max                     3.000000e+01  8.100000e+01          3.000000e+00   

             income  
count  3.243421e+07  
mean   9.943525e+04  
std    4.305473e+04  
min    2.590300e+04  
25%    6.699800e+04  
50%    9.661200e+04  
75%    1.279080e+05  
max    5.939010e+05

In [32]:
# Exporting the new dataframe as a pickle file
ords_prods_custs_merge.to_pickle(os.path.join(path_name, '02_Data', 'Prepared Data', 'ords_prods_custs_merge.pkl'))